File: chap-3/python/stack.py
"""
    Nii Mante
    Stack implementation
"""

from node import Node

class Stack:

    def __init__(self, data):
        self.top = Node(data)


    def __str__(self):
        tmp = self.top
        out_str = "Stack\n"
        out_str = out_str + "====\n"
        while tmp:
            out_str = out_str + str(tmp.data) + "\n" 
            tmp = tmp.next_node

        out_str = out_str + "\n====\n"
        return out_str   

    def pop(self):
        if self.top:
            result = self.top
            self.top = self.top.next_node
            return result

        return None


    def push(self, data):
        new_node = Node(data)
        new_node.next_node = self.top
        self.top = new_node

File: chap-3/cpp/ds/stack.cpp
#include "stack.hpp"

template <class T>
Node<T> *Stack<T>::pop() {
	Node<T> *result = 0;
	if (top){
		result = new Node<T>(top->data, top->next);
		Node<T> *toDelete = top;
		top = top->next; 
		delete toDelete;
		return result;
	}
	return result;
}

template <class T>
T Stack<T>::popData(){
	Node<T> *result = pop();
	if(result){
		return result->getData();
	}
	return 0;
}

template <class T>
void Stack<T>::push(T data) {
	Node<T> *newNode = new Node<T>(data, this->top); 
	top = newNode;
}

template <class T>
T Stack<T>::peek(){
	return top ? top->data : 0; 
}

template <class T>
bool Stack<T>::empty(){
	return top == NULL ? true : false;
}

template <class A>
std::ostream& operator <<(std::ostream &out, const Stack<A> &s){
	Node<A> *t = s.top;
	if(!t){
		out << "Stack empty" << std::endl;
	}else{
		out << "Stack" << std::endl;
		out << " === " << std::endl;
		while(t){
			out << "| " << t->data << std::endl;
			t = t->next;
		}

		out << " === " << std::endl;
	}
	return out;
}
